### ISSUE TITLE
Hyperdiffusion term does not conserve helicity

### ISSUE THREAD

**Issue by vaoskar** [chalmersplasmatheory/DREAM#174]
It appears that a possible reason for this error is that the hyperdiffusion term does not seem to account for the extra factors of V' that enters via the DiffusionTerm. This could be solved by compensating with a factor 1/V' in the diffusion coefficient and re-normalising the equation for dpsi_p/dt so that it actually should be a factor 1/V' in front of the diffusion term.

**Comment by Embreus** [comment]
Disclaimer: my view may be outdated by a year or so.
Which definition of helicity? From the implementation of the psi_p equation (Settings/Equations/E_field.cpp) it is quite clear what the conserved quantity is, since the hyperdiffusion is implemented as a pure diffusion term in the rescaled poloidal-flux equation.

**Comment by hoppe93** [comment]
I think the motivation for Oskar's modification was that the equation currently implemented on master does not appear to be the same as in ``theory.pdf``. In DREAM the hyperdiffusion term is implemented as a pure diffusion term, while it actually should have a somewhat different structure (i.e. the expression inside the outer r-derivative should have a 1/V', instead of the V' set in the diffusion term).

According to István, helicity is not conserved in the current implementation, and supposedly to a somewhat worrisome degree (relative change 1e-3 in a simulation). The situation seems to improve somewhat with Oskar's modification, so that in cylindrical geometry the error is ~1e-5 (which is the same error as obtained without the hyperdiffusion), while in toroidal geometry it is ~1e-4. The error does not appear to depend on resolution (beyond a certain point).

@vaoskar and @pusztai may be able to provide more details, as I have only been peripherally involved in this.

**Comment by pusztai** [comment]
The usual diffusion terms are of the form
1/V' d/dr [V' D d/dr (...)]
and these conserve the quantity they diffuse (that is the same quantity that is being evolved and that appears under the operation). 
The term describing current relaxation is of the form
d/d psi_t [Lambda ... d/d psi_t(...)]
and this conserves magnetic helicity (instead of psi_p that is being evolved, or j/B appearing under the operation).
The problem was that the implemented version had the 1/V' and V' factors in it, as in a usual diffusion term, which needed to be removed. In addition, today we also identified some a misplaced b_min that needed to be fixed (this got rid of the remaining small error at finite aspect ratio). 
The corrected implementation of the psi_p evolution equation is in branch iss174, and I have confirmed that now it conserves magnetic helicity (as much as possible in the presence of a finite resistivity), and in case of Lambda=0, the electric field evolution is the same as before the fix (that is, the V_loop part also does what it is supposed to).
